storage.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 Teserakt AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e4
+
+import (
+	"bytes"
+	"errors"
+	"io"
+)
+
+// Store defines an interface for E4 storage implementation
+// it is compatible with io.ReadWriteSeeker but redefined to allow this type to be exported
+// from bindings generation
+type Store interface {
+	Write(p []byte) (n int, err error)
+	Read(b []byte) (n int, err error)
+	Seek(offset int64, whence int) (idx int64, err error)
+}
+
+type memoryStore struct {
+	buf   *bytes.Buffer
+	index int64
+}
+
+var _ Store = (*memoryStore)(nil)
+
+// NewMemoryStore creates a new Store in memory
+func NewMemoryStore(buf []byte) Store {
+	return &memoryStore{
+		buf: bytes.NewBuffer(buf),
+	}
+}
+
+func (s *memoryStore) Write(p []byte) (n int, err error) {
+	if s.index < 0 {
+		return 0, io.EOF
+	}
+
+	idx := int(s.index)
+	bufLen := s.buf.Len()
+	if idx != bufLen && idx <= bufLen {
+		bufSlice := s.buf.Bytes()[:s.index]
+		s.buf = bytes.NewBuffer(bufSlice)
+	}
+
+	n, err = s.buf.Write(p)
+	s.index += int64(n)
+
+	return n, err
+}
+
+func (s *memoryStore) Read(b []byte) (n int, err error) {
+	if len(b) == 0 {
+		return 0, nil
+	}
+	if s.index >= int64(s.buf.Len()) {
+		return 0, io.EOF
+	}
+
+	bufSlice := s.buf.Bytes()[s.index:]
+	n, err = bytes.NewBuffer(bufSlice).Read(b)
```

Comment:
This is a very expensive way of performing every single read. [10]byte[:] read will involve reslicing the whole underlying buffer. I think just using bytes.Buffer will do the job and it already handle io.EOF if the length is hit.

Reply:
I'm not sure to understand here, can you describe a bit more what you mean ?

Reply:
You are reslicing the underlying byteslice and creating a new bytes.Buffer on every read. That means for example:
```go
_1kBuf := make([]bye, 1<<10)
for {
    n, err := store.Read(buf)
    if err != io.EOF {…}
    // Use buf
    _ = buf[:n]
}
```

Or io.Copy(dst, store)

Will reslice the original underlying buffer and re-create bytes.Buffer as many times as Read is invoked and will be very CPU and memory intensive.

bytes.Buffer is an already an io.Reader as well as an io.Writer, or is the problem just knowing where the head should be? If so, bytes.Reader is already an io.ReadWriteSeeker and can eliminate all the logic which is error prone, of which this new logic will need lots of edge case testing https://golang.org/pkg/bytes/#Reader.

I’d say bytes.Reader could literally be in the constructor for NewInMemoryStore and no other code will be needed.